client/components/credit-card-form-fields/index.jsx

```diff
@@ -17,6 +17,7 @@ import CreditCardNumberInput from 'components/upgrades/credit-card-number-input'
 import PaymentCountrySelect from 'components/payment-country-select';
 import EbanxPaymentFields from 'my-sites/checkout/checkout/ebanx-payment-fields';
 import { Input } from 'my-sites/domains/components/form';
+import InfoPopover from 'components/info-popover';
```

Comment:
This component assumes window is available, so test setup for `client/my-sites/checkout/checkout/test/checkout.js` should be updated. Please don't merge with this broken unit test

```jsx
Summary of all failing tests
 FAIL  client/my-sites/checkout/checkout/test/checkout.js
  ● Test suite failed to run

    ReferenceError: window is not defined

      34 | 
      35 | class Popover extends Component {
    > 36 |     static propTypes = {
         |                                   ^
      37 |         autoPosition: PropTypes.bool,
      38 |         autoRtl: PropTypes.bool,
      39 |         className: PropTypes.string,

      at Object.<anonymous> (node_modules/component-event/index.js:1:39)
      at Object.<anonymous> (node_modules/click-outside/build/index.js:13:23)
      at Object.<anonymous> (client/components/popover/index.jsx:36:44)
```

Reply:
To quickly run this test `npm run test-client client/my-sites/checkout/checkout/test/checkout.js`

We should see the following output:
```
> wp-calypso@0.17.0 pretest-client /Users/kerryliu/checkouts/wp-calypso
> npm run -s pretest


> wp-calypso@0.17.0 test-client /Users/kerryliu/checkouts/wp-calypso
> jest -c=test/client/jest.config.json "client/my-sites/checkout/checkout/test/checkout.js"

 FAIL  client/my-sites/checkout/checkout/test/checkout.js
  ● Test suite failed to run

    ReferenceError: window is not defined

      10 | import debugFactory from 'debug';
      11 | import classNames from 'classnames';
    > 12 | import clickOutside from 'click-outside';
         | ^
      13 | import { uniqueId } from 'lodash';
      14 | 
      15 | /**

      at Object.<anonymous> (node_modules/component-event/index.js:1:39)
      at Object.<anonymous> (node_modules/click-outside/index.js:6:1)
      at Object.<anonymous> (client/components/popover/index.jsx:12:1)

Test Suites: 1 failed, 1 total
Tests:       0 total
Snapshots:   0 total
Time:        11.131s
Ran all test suites matching /client\/my-sites\/checkout\/checkout\/test\/checkout.js/i.
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! wp-calypso@0.17.0 test-client: `jest -c=test/client/jest.config.json "client/my-sites/checkout/checkout/test/checkout.js"`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the wp-calypso@0.17.0 test-client script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/kerryliu/.npm/_logs/2018-07-25T16_35_53_295Z-debug.log
```

Reply:
Let's see, the easiest thing to do is update the `client/my-sites/checkout/checkout/test/checkout.js` file to note that it needs jsdom. It'll be a bit noisy though since this apparently doesn't mock scrollTo silently
```
/**
 * @format
 * @jest-environment jsdom
 */
```
I'm seeing output like so. Any suggestions @gziolo @jsnajdr ?

```
$ npm run test-client client/my-sites/checkout/checkout/test/checkout.js

> wp-calypso@0.17.0 pretest-client /Users/kerryliu/checkouts/wp-calypso
> npm run -s pretest


> wp-calypso@0.17.0 test-client /Users/kerryliu/checkouts/wp-calypso
> jest -c=test/client/jest.config.json "client/my-sites/checkout/checkout/test/checkout.js"

 PASS  client/my-sites/checkout/checkout/test/checkout.js
  ● Console

    console.error node_modules/jsdom/lib/jsdom/virtual-console.js:29
      Error: Not implemented: window.scrollTo
          at module.exports (/Users/kerryliu/checkouts/wp-calypso/node_modules/jsdom/lib/jsdom/browser/not-implemented.js:9:17)
          at /Users/kerryliu/checkouts/wp-calypso/node_modules/jsdom/lib/jsdom/browser/Window.js:544:7
          at Checkout.componentDidMount (/Users/kerryliu/checkouts/wp-calypso/client/my-sites/checkout/checkout/checkout.jsx:104:10)
          at /Users/kerryliu/checkouts/wp-calypso/node_modules/enzyme/build/ShallowWrapper.js:126:20
          at Object.batchedUpdates (/Users/kerryliu/checkouts/wp-calypso/node_modules/enzyme-adapter-react-16/build/ReactSixteenAdapter.js:342:22)
          at new ShallowWrapper (/Users/kerryliu/checkouts/wp-calypso/node_modules/enzyme/build/ShallowWrapper.js:125:24)
          at shallow (/Users/kerryliu/checkouts/wp-calypso/node_modules/enzyme/build/shallow.js:19:10)
          at Object.<anonymous> (/Users/kerryliu/checkouts/wp-calypso/client/my-sites/checkout/checkout/test/checkout.js:81:20)
          at Object.asyncFn (/Users/kerryliu/checkouts/wp-calypso/node_modules/jest-jasmine2/build/jasmine_async.js:108:37)
          at resolve (/Users/kerryliu/checkouts/wp-calypso/node_modules/jest-jasmine2/build/queue_runner.js:56:12) undefined
    console.error node_modules/jsdom/lib/jsdom/virtual-console.js:29
      Error: Not implemented: window.scrollTo
          at module.exports (/Users/kerryliu/checkouts/wp-calypso/node_modules/jsdom/lib/jsdom/browser/not-implemented.js:9:17)
          at /Users/kerryliu/checkouts/wp-calypso/node_modules/jsdom/lib/jsdom/browser/Window.js:544:7
          at Checkout.componentDidMount (/Users/kerryliu/checkouts/wp-calypso/client/my-sites/checkout/checkout/checkout.jsx:104:10)
          at /Users/kerryliu/checkouts/wp-calypso/node_modules/enzyme/build/ShallowWrapper.js:126:20
          at Object.batchedUpdates (/Users/kerryliu/checkouts/wp-calypso/node_modules/enzyme-adapter-react-16/build/ReactSixteenAdapter.js:342:22)
          at new ShallowWrapper (/Users/kerryliu/checkouts/wp-calypso/node_modules/enzyme/build/ShallowWrapper.js:125:24)
          at shallow (/Users/kerryliu/checkouts/wp-calypso/node_modules/enzyme/build/shallow.js:19:10)
          at Object.<anonymous> (/Users/kerryliu/checkouts/wp-calypso/client/my-sites/checkout/checkout/test/checkout.js:88:14)
          at Object.asyncFn (/Users/kerryliu/checkouts/wp-calypso/node_modules/jest-jasmine2/build/jasmine_async.js:108:37)
          at resolve (/Users/kerryliu/checkouts/wp-calypso/node_modules/jest-jasmine2/build/queue_runner.js:56:12) undefined
    console.error node_modules/jsdom/lib/jsdom/virtual-console.js:29
      Error: Not implemented: window.scrollTo
          at module.exports (/Users/kerryliu/checkouts/wp-calypso/node_modules/jsdom/lib/jsdom/browser/not-implemented.js:9:17)
          at /Users/kerryliu/checkouts/wp-calypso/node_modules/jsdom/lib/jsdom/browser/Window.js:544:7
          at Checkout.componentDidMount (/Users/kerryliu/checkouts/wp-calypso/client/my-sites/checkout/checkout/checkout.jsx:104:10)
          at /Users/kerryliu/checkouts/wp-calypso/node_modules/enzyme/build/ShallowWrapper.js:126:20
          at Object.batchedUpdates (/Users/kerryliu/checkouts/wp-calypso/node_modules/enzyme-adapter-react-16/build/ReactSixteenAdapter.js:342:22)
          at new ShallowWrapper (/Users/kerryliu/checkouts/wp-calypso/node_modules/enzyme/build/ShallowWrapper.js:125:24)
          at shallow (/Users/kerryliu/checkouts/wp-calypso/node_modules/enzyme/build/shallow.js:19:10)
          at Object.<anonymous> (/Users/kerryliu/checkouts/wp-calypso/client/my-sites/checkout/checkout/test/checkout.js:100:20)
          at Object.asyncFn (/Users/kerryliu/checkouts/wp-calypso/node_modules/jest-jasmine2/build/jasmine_async.js:108:37)
          at resolve (/Users/kerryliu/checkouts/wp-calypso/node_modules/jest-jasmine2/build/queue_runner.js:56:12) undefined
    console.error node_modules/jsdom/lib/jsdom/virtual-console.js:29
      Error: Not implemented: window.scrollTo
          at module.exports (/Users/kerryliu/checkouts/wp-calypso/node_modules/jsdom/lib/jsdom/browser/not-implemented.js:9:17)
          at /Users/kerryliu/checkouts/wp-calypso/node_modules/jsdom/lib/jsdom/browser/Window.js:544:7
          at Checkout.componentDidMount (/Users/kerryliu/checkouts/wp-calypso/client/my-sites/checkout/checkout/checkout.jsx:104:10)
          at /Users/kerryliu/checkouts/wp-calypso/node_modules/enzyme/build/ShallowWrapper.js:126:20
          at Object.batchedUpdates (/Users/kerryliu/checkouts/wp-calypso/node_modules/enzyme-adapter-react-16/build/ReactSixteenAdapter.js:342:22)
          at new ShallowWrapper (/Users/kerryliu/checkouts/wp-calypso/node_modules/enzyme/build/ShallowWrapper.js:125:24)
          at shallow (/Users/kerryliu/checkouts/wp-calypso/node_modules/enzyme/build/shallow.js:19:10)
          at Object.<anonymous> (/Users/kerryliu/checkouts/wp-calypso/client/my-sites/checkout/checkout/test/checkout.js:110:20)
          at Object.asyncFn (/Users/kerryliu/checkouts/wp-calypso/node_modules/jest-jasmine2/build/jasmine_async.js:108:37)
          at resolve (/Users/kerryliu/checkouts/wp-calypso/node_modules/jest-jasmine2/build/queue_runner.js:56:12) undefined
```

Reply:
To fix the test in `client/my-sites/checkout/checkout/test/checkout.js`
Let's add the following to the top of the file. (Adds the jsdom module)
```
/**
 * @format
 * @jest-environment jsdom
 */
```

And to avoid the console.error logs, near the jest.mock calls we can also add:
```
//jsdom doesn't properly mock scrollTo
global.scrollTo = () => {};
```

I think `jest` is smart enough to encapsulate this, so this doesn't pollute other tests. (Normally we'd do something like this in test setup/teardown)

Reply:
Yes, jsdom and mocking missing global is the way to go. Those changes will be isolated for the file only. Jest will take care of everything.

Reply:
Related section in Jest docs: https://jestjs.io/docs/en/manual-mocks#mocking-methods-which-are-not-implemented-in-jsdom